controllers/managedosversion_controller.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright © 2022 - 2024 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	elementalv1 "github.com/rancher/elemental-operator/api/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ManagedOSVersionReconciler reconciles a ManagedOSVersion object.
+type ManagedOSVersionReconciler struct {
+	client.Client
+	DefaultRegistry string
+	Scheme          *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=elemental.cattle.io,resources=managedosversions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=elemental.cattle.io,resources=managedosversions/status,verbs=get;update;patch;list
+// +kubebuilder:rbac:groups=elemental.cattle.io,resources=managedosimages,verbs=get;list;watch
+
+func (r *ManagedOSVersionReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&elementalv1.ManagedOSVersion{}).
+		Watches(
+			&elementalv1.ManagedOSImage{},
+			handler.EnqueueRequestsFromMapFunc(r.ManagedOSImageToManagedOSVersion),
+		).
+		Complete(r)
+}
+
+func (r *ManagedOSVersionReconciler) ManagedOSImageToManagedOSVersion(ctx context.Context, obj client.Object) []ctrl.Request {
+	logger := ctrl.LoggerFrom(ctx)
+	logger.Info("Enqueueing ManagedOSVersion reconciliation from ManagedOSImage")
+
+	requests := []ctrl.Request{}
+
+	// Verify we are actually handling a ManagedOSImage object
+	managedOSImage, ok := obj.(*elementalv1.ManagedOSImage)
+	if !ok {
+		logger.Error(errors.New("Enqueueing error"), fmt.Sprintf("Expected a ElementalHost object, but got %T", obj))
+		return []ctrl.Request{}
+	}
+
+	// Check the ManagedOSImage was referencing any ManagedOSVersion
+	if len(managedOSImage.Spec.ManagedOSVersionName) > 0 {
```

Comment:
🤔 is this to clean-up stale ManagedOSVersion resources when the ManagedOSImage is deleted?

Reply:
Yes. 

The ManagedOSVersionController will **not** requeue after trying to remove the finalizer the first time, so we need to watch ManagedOSImages to check for changes. Makes it pretty efficient.